### PR TITLE
476 Add and update type hints

### PR DESCRIPTION
**Pull request by benjamin-gorman**
### Description
Another iteration of adding missing type hints.

Type hint updates
- [`Union[int, float] ≡ float`](https://www.python.org/dev/peps/pep-0484/#the-numeric-tower)
- ~~`val_handlers: Optional[Sequence[Metric]]`, previously `Optional[Sequence]`~~
- ~~`loss_fn: torch.nn.modules.loss._Loss`, previously `Callable`~~
- `dtype: np.dtype`, previously `Optional[np.dtype]`
  - to match underlying function call

### Status
**Ready**

### Types of changes
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.

### CODE REVIEW

File: monai/data/nifti_reader.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def __init__(
         self,
         image_files: Sequence[str],
         seg_files: Optional[Sequence[str]] = None,
-        labels: Optional[Sequence[Union[int, float]]] = None,
+        labels: Optional[Sequence[float]] = None,
```
Comment (Nic-Ma) [inline monai/data/nifti_reader.py:31]: May I know why you change from `Union[int, float]` to `float`?

File: monai/engines/evaluator.py [hunk 2]
```diff
@@ -59,7 +59,7 @@ def __init__(
         post_transform: Optional[Transform] = None,
         key_val_metric: Optional[Dict[str, Metric]] = None,
         additional_metrics: Optional[Dict[str, Metric]] = None,
-        val_handlers: Optional[Sequence] = None,
+        val_handlers: Optional[Sequence[Metric]] = None,
```
Comment (Nic-Ma) [inline monai/engines/evaluator.py:62]: val_handler is not `Metric`, it's MONAI own classes under `handlers/`, basically, it can be any object that can handle data, so I just put `Sequence` for it.

File: monai/engines/evaluator.py [hunk 3]
```diff
@@ -126,7 +126,7 @@ def __init__(
...
```
Comment (Nic-Ma) [inline monai/engines/evaluator.py:129]: same as above.

File: monai/engines/trainer.py [hunk 4]
```diff
@@ -84,15 +84,15 @@ def __init__(
...
+        loss_function: torch.nn.modules.loss._Loss,
         prepare_batch: Callable = default_prepare_batch,
         iteration_update: Optional[Callable] = None,
         inferer: Inferer = SimpleInferer(),
         amp: bool = True,
         post_transform: Optional[Transform] = None,
         key_train_metric: Optional[Dict[str, Metric]] = None,
         additional_metrics: Optional[Dict[str, Metric]] = None,
-        train_handlers: Optional[Sequence] = None,
+        train_handlers: Optional[Sequence[Metric]] = None,
```
Comment (Nic-Ma) [inline monai/engines/trainer.py:95]: same as `val_handlers`.

File: monai/engines/workflow.py [hunk 5]
```diff
@@ -69,7 +69,7 @@ def __init__(
         post_transform: Optional[Callable] = None,
         key_metric: Optional[Dict[str, Metric]] = None,
         additional_metrics: Optional[Dict[str, Metric]] = None,
-        handlers: Optional[Sequence] = None,
+        handlers: Optional[Sequence[Metric]] = None,
```
Comment (Nic-Ma) [inline monai/engines/workflow.py:72]: same as `val_handlers`.

File: monai/engines/multi_gpu_supervised_trainer.py [hunk 6]
```diff
@@ -37,7 +37,7 @@ def _default_eval_transform(x, y, y_pred):
 def create_multigpu_supervised_trainer(
     net: torch.nn.Module,
     optimizer: Optimizer,
-    loss_fn: Callable,
+    loss_fn: torch.nn.modules.loss._Loss,
```
Comment (ericspod) [inline monai/engines/multi_gpu_supervised_trainer.py:40]: I would suggest changing this back to `Callable`. There's no need for the loss function to be a Module type at all, it could be any callable which returns a tensor with gradients, that's all that's needed. I've seen loss functions implemented as Python routines before, I'm sure there's other loss functions implemented as direct subclasses of `nn.Module` too.

File: monai/engines/trainer.py [hunk 7]
```diff
@@ -84,15 +84,15 @@ def __init__(
         train_data_loader: DataLoader,
         network: torch.nn.Module,
         optimizer: Optimizer,
-        loss_function: Callable,
+        loss_function: torch.nn.modules.loss._Loss,
```
Comment (ericspod) [inline monai/engines/trainer.py:87]: Same here as previous comment.

File: monai/engines/multi_gpu_supervised_trainer.py (hunk 6 above)
Comment (ericspod) [reply monai/engines/multi_gpu_supervised_trainer.py:40]: Apologies if this has already been discussed, it's something that caught my attention here.

File: monai/data/nifti_reader.py (hunk 1 above)
Comment (benjamin-gorman) [reply monai/data/nifti_reader.py:31]: https://www.python.org/dev/peps/pep-0484/#the-numeric-tower

> Rather than requiring that users write import numbers and then use numbers.Float etc., this PEP proposes a straightforward shortcut that is almost as effective: when an argument is annotated as having type float, an argument of type int is acceptable

File: monai/engines/evaluator.py (hunk 2 above)
Comment (benjamin-gorman) [reply monai/engines/evaluator.py:62]: My bad, I was constraining it for this call. I should look into some sort of duck typing solution.

```python
handler.attach(self)
```

File: monai/engines/multi_gpu_supervised_trainer.py (hunk 6 above)
Comment (benjamin-gorman) [reply monai/engines/multi_gpu_supervised_trainer.py:40]: My bad, I had my focus too much on torch losses; good to know.

**Comment by ericspod** [comment]
No, looks good to me.
